docs/signature-layout.md

```diff
@@ -0,0 +1,35 @@
+# Signature File Layout
+
+A common file layout for storing and serving signatures provides a consistent way to reference image signatures. Signatures on a filesystem or a web server shall use this common layout. Signatures stored in a REST API are not required to use this common layout.
+
+## Specification
+
+This specification relies on [RFC3986](https://tools.ietf.org/html/rfc3986), focusing on defining a [path component](https://tools.ietf.org/html/rfc3986#section-3.3) to compose a concise URI reference to a signature.
+
+**SCHEME[AUTHORITY]/PATH_PREFIX/IMAGE@MANIFEST_DIGEST/signature-INT**
+
+**Definitions**
+
+* **SCHEME**: URI scheme per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.1), e.g. **file://** or **https://**
+* **AUTHORITY**: An optional authority reference per [RFC3986](https://tools.ietf.org/html/rfc3986#section-3.2), e.g. **example.com**
+* **PATH_PREFIX**: An arbitrary base path to the image component
+* **IMAGE**: The name of the image per [v2 API](https://docs.docker.com/registry/spec/api/#/overview). This would typically take the form of registry/repository/image but is not required to have exactly three parts. There is no requirement to include a **:PORT** component but it should be included if part of the image reference.
```

Comment:
> Attaching signatures to manifests makes the logistics of signature distribution MUCH easier. However, per #59, we are discussing "a detached container image signature object and signing methods".

\*shrug\* I don’t think attached/detached in #59 was supposed to mean _that_. It is detached in the sense of “detached GPG signature”; an extra file which does not modify the signed content. It is not detached in the sense of “floating somewhere on the internet in a distributed hash table or on the block chain, find it if you can”.

> If we instead are going to discuss an attached model, that would be entirely different. I like where you're going, thinking about signatures as an additional entity managed by a registry, side-by-side with manifests in the same namespace. But at this point it does not seem realistic to get it added to a registry protocol, or else we should be having this discussion elsewhere.

This is actually already supported in the OpenShift embedded registry. (Well, the client APIs to get images and signatures are kind of different, but the underlying storage and implementation is truly shared.)

> As such, I don't see how we can talk about this as "an invisible implementation detail simulating real registries with native signature support". How do we simulate something that doesn't exist?

A significant value provided by `containers/image` is that it tries to give a uniform interface for various image sources/destinations, and universal tools like `copy.Image` (aka `skopeo copy`) for work with, as much as possible, any combinations of sources/destinations. That, of course, only works if their behavior is similar enough.

The semantics implied by `containers/image/types.Image{Source,Destination}` is that signatures are a part of a single image just as the blob layers are, not referenced separately, and the user uses a single string (`dir:$path`, `docker://$docker-reference`) to refer to the image as a whole, manifest, layers, signatures, everything.

For OpenShift, this happens natively; for docker/distribution registries, we need this sigstore crutch at the moment, and the user needs to perform an one-time setup to configure the sigstore URLs. Yes, we do find it practical to the sigstore roots to be configurable at both more detailed granularity (individual sigstore roots for individual repos) and a more general granularity (a single default sigstore for all hosts), for various reasons; but as far as the rest of `containers/image` cares, these capabilities of `registries.d/*.yaml` don’t exist and the signatures are an inseparable part of every single image.

> We could consider requiring that a sigstore be tightly coupled to a specific registry. When you deploy a registry, you deploy a sigstore with it whose namespaces will match those in the registry.

Yes, I would expect this to be the most natural case: a single sigstore for a single docker/distribution registry. There are other possibilities and it might not even be the most _common_ case, but it is the most _natural_ one.

> A problem though is that every tool that could move images around, from one registry to another, must also be aware of accompanying sigstores and move signatures together with images.

Yes, _and that is unavoidable_. Sigstores, like registries, are necessarily federated (various ISVs will host their own registries, possibly open only to paying customers, and nobody else would care to host sigstores for their private contents just for fun). So, every tool moving images around between such separately-managed repositories needs to be aware of the accompanying sigstores; or at least aware enough to use something which understands the system-wide configuration in `registries.d/*.yaml`.

> Now we would need a standard API for adding signatures to a sigstore and copying them within a sigstore. Every tool that manipulates a registry would need to use that API to manipulate the accompanying sigstore.

That exists already: `containers/image` and `skopeo`.  Use the `containers/image` image reference format for all images, wherever they are, and then just use `skopeo copy` or call `containers/image/copy.Image`.  `skopeo copy` will read `registries.d/*.yaml` and interpret it properly, copying images + signatures from one place to another, _even across various registry mechanisms_ (between docker/distribution+sigstore and OpenShift, and whatever else will be invented in the future.)

> The problem not solved by the namespaced-sigstore model, as this PR defines it, is one of mirroring. Consider that I have an on-premises registry at reg.hrivnak.org. I use it to have a local mirror of third-part content, so it includes repositories like redhat/rhel7. A local client tries to pull reg.hrivnak.org/redhat/rhel7:latest and gets some content. At what URL should it expect to find a signature?

An unconfigured client would not have a sigstore configuration for `reg.hrivnak.org` and it will not fetch any signatures.

Again, the conceptual model is that image+signatures are a single unit always hosted together. So, if you have a docker/distribution registry mirror, you should also have a corresponding sigstore mirror, and the clients need to be configured to know the sigstore location for `reg-sigstore.hrivnak.org`.

> A flat directory of signatures would easily solve this issue. It has disadvantages, but it's strength is this: because a manifest is not inherently tied to a specific registry, its signatures also should not be tied to a specific registry.

It wouldn’t; how flat would this directory have to be? AFAICT for your case to work, and all signatures to be automatically found for whatever registry anyone in the world decides to host, there would need to be a _single world-wide sigstore_. Who would host it? Why? Would everyone want to make their images public like that? Would they want to cede control to that host? Essentially this would recreate the “Docker Hub as a single point of failure/control” design again, with all its issues.

> Access control with a flat namespace could be reasonably achieved with multiple sigstores. If you have a secret key that you use to sign things for secret purposes, put those signatures in a secret/protected sigstore. This allows you to orient a sigstore with respect to an authority and an audience, which is primarily how content gets consumed. As a user, I want to get signatures from a particular authority that I trust. As an authority, I want to make my signatures available to a defined set of users. Thus the practical relationship between a signing authority and its audience may not need to include registry names or repository namespaces.

See the ACME example above; it seems to me far more natural, and it saves a lot of administrative overhead, to have a single large registry with granular access control, possibly even multi-tenant, than a number of independent all-or-nothing registries with all-or-nothing access control

> @mtrmac I have a feeling that you have a particular scenario in mind that involves many users, many keys, and lots of access restrictions. Perhaps it would help to define those use cases directly and explicitly, and then we could more easily see which designs would accommodate them.

I think the conceptual hurdle is to ignore how the sigstore technically works and to see that containers/image generally treats manifest+layers+signatures as _single_ atomic unit. Then the philosophy and access control issues etc. follow naturally.

> I am not at all tied to the idea of a flat namespace, but I am very interested in covering use cases that involve mirroring.

A mirror of docker/distribution+sigstore will necessarily have to be a pair of docker/distribution+sigstore servers; otherwise it would only be a partial mirror.

(FWIW note that naive mirroring signatures with names cannot “just work” because the signatures claim the original (`redhat/rhel7`) identities, so pulling the  `reg.hrivnak.org/redhat/rhel7:latest` will see an identity mismatch. That is quite unrelated to the sigstore mechanism, and we are prepared to support that, see `hostname:5000/vendor/product` in https://github.com/containers/image/blob/master/docs/policy.json.md#examples .)